fix(input): allow overriding timezone for date input types

Narretz opened the pull request:
Fixes #13382
Closes #16181

<!-- General PR submission guidelines https://github.com/angular/angular.js/CONTRIBUTING.md#submit-pr -->
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
fix


**What is the current behavior? (You can also link to an open issue here)**
See https://github.com/angular/angular.js/issues/16181


**What is the new behavior (if this is a feature change)?**
overrideModelOptions can be used to change the timezone.


**Does this PR introduce a breaking change?**
No


**Please check if the PR fulfills these requirements**
- [x] The commit message follows our [guidelines](../DEVELOPERS.md#commits)
- [x] Fix/Feature: [Docs](../DEVELOPERS.md#documentation) have been added/updated
- [x] Fix/Feature: Tests have been added; existing tests pass

**Other information**:

Narretz committed:
fix(input): allow overriding timezone for date input types

This commit also fixes a bug where part of the Date object
was re-used even after the input was emptied.

Fixes #16181
Closes #13382

petebacondarwin reviewed (approved):
nice work @Narretz

gkalpak commented on test/ng/directive/inputSpec.js:
```diff
@@ -662,6 +662,21 @@ describe('input', function() {
+      inputElm.controller('ngModel').$overrideModelOptions({timezone: '-0500'});
+
+      $rootScope.$apply(function() {
+        $rootScope.value = new Date(Date.UTC(2014, 6, 1));
```
Why the year change (here and below)? I would stick to the minimum necessary changes to avoid confusion.

gkalpak reviewed (commented):
I wonder if we should document that changing the value timezone will have no effect until the next parse/format.

Narretz commented on src/ng/directive/input.js:
```diff
@@ -1431,17 +1431,20 @@ function createDateInputType(type, regexp, parseDate, format) {
     ctrl.$parsers.push(function(value) {
-      if (ctrl.$isEmpty(value)) return null;
+      if (ctrl.$isEmpty(value)) {
+        previousDate = null;
```
If it's not valid, then why are there no test failures? :D
The formatter is doing the same thing afaict - it sets previousDate = null if the model is not a valid date (empty is not a valid date, too).
The specific test at https://github.com/angular/angular.js/pull/16336/files/d172e8ecb650ad3220e00e46b08325ad3672e9f1#diff-ce6d0f141f60d25eede7abf044d04b93R1968 shows that when clearing out the input, previousDate is still set and this leads the next input to use an incorrect value for the hour part.

We'd need a test case to show that this is a regression.

gkalpak replied:
> If it's not valid, then why are there no test failures? :D

Valid point...nooot 😛 

> The formatter is doing the same thing afaict - it sets `previousDate = null` if the model is not a valid date (empty is not a valid date, too).

That is the point. The formatter processes the model value and `previousDate` keeps track of the last valid (date) model value.

> The specific test at https://github.com/angular/angular.js/pull/16336/files/d172e8ecb650ad3220e00e46b08325ad3672e9f1#diff-ce6d0f141f60d25eede7abf044d04b93R1968 shows that when clearing out the input, previousDate is still set and this leads the next input to use an incorrect value for the hour part.

"use an incorrect value for the hour part" is subjective. I'd say currently it uses "the correct value for the hour part".
The idea is to be able to set the hour for a date field. AFAICT, the only way to achieve this today is by setting the model to a date with the desired date. Then, as the user edits the field (including emptying and refilling it) the hour/minute/second part is preserved. I think this is a desired property (I am pretty sure I've used it in my apps :smiley:).
In any case, I believe this change is breaking not related to this PR. It should go into a separate PR (where it can be debated further :grin:).

Narretz replied:
> The idea is to be able to set the hour for a date field.

Ah, that's right. Did you get this from the code or did you search for the commit? Here's the commit: https://github.com/angular/angular.js/commit/1a1ef62903c8fdf4ceb81277d966a8eff67f0a96

As you can see, there are not tests for type=date ... :<

This is at least confusing, because as you can see in the test, the date is re-using the hour part that was set for -0500, even though the timezone is set to UTC again. But maybe this is a contrived example.

gkalpak replied:
I didn't search for the commit (I didn't expect there to be a commit specific for that - I assumed it was part of the initial implementation) :smiley:

So, I am not sure what the correct behavior is here (wrt to changing the timezone) 😕
Maybe keep track of the previous timezone and re-adjust? Or accept the fact that changing the timezone may change the time part?
Timezones are tricky :grin:

Narretz commented:
Regarding the docs - any property changes in ngModelOptions only have an effect after the next update. We could add this info in overrideModelOptions generally.

Narretz committed:
fixup! remove incorrect date reset bugfix

Narretz commented:
I've removed the date reset and added a different fix that should account for changing timezones. I'm not 100% happy with the implementation, but I also don't want to spend too much time on this.

Narretz commented:
@gkalpak could you give this another look?

gkalpak commented on test/ng/directive/inputSpec.js:
```diff
@@ -931,6 +946,22 @@ describe('input', function() {
+      helper.changeInputValueTo('2013-W03');
+      expect(+$rootScope.value).toBe(Date.UTC(2013, 0, 17));
+
+      inputElm.controller('ngModel').$overrideModelOptions({timezone: '+5000'});
```
Is this really supposed to be `+5000`?

Narretz replied:
Revised the test and added more comments.

gkalpak commented on test/ng/directive/inputSpec.js:
```diff
@@ -931,6 +946,22 @@ describe('input', function() {
+      inputElm.controller('ngModel').$overrideModelOptions({timezone: '+5000'});
+
+      $rootScope.$apply(function() {
+        // the 17. with an offset of +5000 moves the date into next week
```
Can you expand a bit on this comment? It mentions `17.` (which is in German format 😛), but the code uses `18`. Also, how does it move to the next week?

gkalpak commented on test/ng/directive/inputSpec.js:
```diff
@@ -1482,6 +1532,25 @@ describe('input', function() {
+
+      inputElm.controller('ngModel').$overrideModelOptions({timezone: 'UTC'});
+      helper.changeInputValueTo('23:02:00');
+      // The year is still set from the previous date
```
👍

gkalpak commented on test/ng/directive/inputSpec.js:
```diff
@@ -1879,6 +1966,30 @@ describe('input', function() {
       dealoc(formElm);
     });
 
+    it('should not reuse the hour part of a previous date object after changing the timezone', function() {
```
Not sure how this test checks that the hour part of the previous date object isn't used.

Narretz replied:
Also Revised this test and added more comments.

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1431,17 +1431,27 @@ function createDateInputType(type, regexp, parseDate, format) {
         // parser/formatter in the processing chain so that the model
         // contains some different data format!
+        if (timezone && previousTimezone && previousTimezone !== timezone) {
+          // If the timezone has changed, adjust the previousDate to the default timzeone
```
_timzeone_ --> _timezone_

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1431,17 +1431,27 @@ function createDateInputType(type, regexp, parseDate, format) {
         // Note: We cannot read ctrl.$modelValue, as there might be a different
         // parser/formatter in the processing chain so that the model
         // contains some different data format!
+        if (timezone && previousTimezone && previousTimezone !== timezone) {
```
Why do we need `timezone` or `previousTimezone` to be truthy? Don't we always need to adjust as long as `previousTimezone !== timezone`?

Narretz replied:
We do need previousTimezone to be defined, since we pass it to addDateMinutes. timezone must not be defined, though.

Narretz committed:
fixup! simplify and improve test info

Narretz reviewed (commented):
I've revised the tests in hope that they are clearer now, or at least it's easier to understand them. Let me know if I have failed :D

gkalpak commented on test/ng/directive/inputSpec.js:
```diff
@@ -949,15 +949,23 @@ describe('input', function() {
 
+      //To check that the timezone overwrite works, apply an offset of +24 hours.
+      //Since January 19 is a Saturday, +24 will turn the formatted Date into January 20 - Sunday -
+      //which is in calendar week 4 instead of 3.
```
Missing whitspace after `//` 😇

gkalpak commented on test/ng/directive/inputSpec.js:
```diff
@@ -1966,28 +1974,32 @@ describe('input', function() {
-
-      // Emptying the input should clear the cached date object
-      helper.changeInputValueTo('');
+      // At this point, the cached Date has its hours set to to 20 (00:00 - 05:00 = 19:00)
```
I am confused 😕
Are you saying that 19 === 20?

Narretz replied:
No, just a result of trying to understand timezones. If you log a date, you always get it with the local timezone offset which is +1 for me, which means 20. But 19 is correct.
